k2/csrc/fsa_algo.cu

```diff
@@ -288,4 +289,92 @@ void ArcSort(Fsa &src, Fsa *dest, Array1<int32_t> *arc_map /*= nullptr*/) {
+
+  bool log_semiring = false;
+  Array1<int32_t> tmp_entering_arcs;
+  Array1<float> forward_scores =
+      GetForwardScores<float>(fsas, state_batches, entering_arc_batches,
+                              log_semiring, &tmp_entering_arcs);
+  if (total_scores != nullptr)
+    *total_scores = GetTotScores<float>(fsas, forward_scores);
+
+  if (entering_arcs != nullptr) *entering_arcs = tmp_entering_arcs;
+
+  const int32_t *entering_arcs_data = tmp_entering_arcs.Data();
+  const Arc *arcs_data = fsas.values.Data();
+  int32_t num_fsas = fsas.Dim0();
+  int32_t num_states = fsas.TotSize(1);
+  ContextPtr &context = fsas.Context();
+
+  // allocate an extra element for ExclusiveSum
+  Array1<int32_t> num_best_arcs_per_fsa(context, num_fsas + 1);
+  int32_t *num_best_arcs_per_fsa_data = num_best_arcs_per_fsa.Data();
+  const int32_t *row_splits1_data = fsas.RowSplits(1).Data();
+
+  // -1 represents an invalid arc_index.
+  // This extra array avoids an extra iteration over `entering_arcs`.
+  Array1<int32_t> state_best_arc_index_array(context, num_states, -1);
+  int32_t *state_best_arc_index_array_data = state_best_arc_index_array.Data();
+
+  auto lambda_set_num_best_arcs = [=] __host__ __device__(int32_t fsas_idx0) {
+    int32_t state_idx01 = row_splits1_data[fsas_idx0];
+    int32_t state_idx01_next = row_splits1_data[fsas_idx0 + 1];
+
+    if (state_idx01_next == state_idx01) {
+      // this fsa is empty, so there is no best path available
+      num_best_arcs_per_fsa_data[fsas_idx0] = 0;
+      return;
+    }
+
+    int32_t final_state_idx01 = state_idx01_next - 1;
+    int32_t cur_state = final_state_idx01;
+    int32_t cur_index = entering_arcs_data[cur_state];
+    int32_t num_arcs = 0;
+    while (cur_index != -1) {
+      state_best_arc_index_array_data[cur_state] = cur_index;
+
+      cur_state = arcs_data[cur_index].src_state + state_idx01;
+      cur_index = entering_arcs_data[cur_state];
+      ++num_arcs;
+    }
+    num_best_arcs_per_fsa_data[fsas_idx0] = num_arcs;
+  };
+  Eval(context, num_fsas, lambda_set_num_best_arcs);
+  ExclusiveSum(num_best_arcs_per_fsa, &num_best_arcs_per_fsa);
+
+  RaggedShape shape = RaggedShape2(&num_best_arcs_per_fsa, nullptr, -1);
+  const int32_t *ans_row_splits_data = shape.RowSplits(1).Data();
+  Array1<int32_t> best_path_arc_indexes(context, shape.NumElements());
+  int32_t *best_path_arc_indexes_data = best_path_arc_indexes.Data();
+
+  auto lambda_set_best_arcs = [=] __host__ __device__(int32_t fsas_idx0) {
+    int32_t state_idx01 = row_splits1_data[fsas_idx0];
+    int32_t state_idx01_next = row_splits1_data[fsas_idx0 + 1];
+    if (state_idx01_next == state_idx01) return;
+
+    int32_t ans_idx01 = ans_row_splits_data[fsas_idx0];
+    int32_t *p = best_path_arc_indexes_data + ans_idx01;
+    for (int32_t s = state_idx01; s != state_idx01_next; ++s) {
+      int32_t index = state_best_arc_index_array_data[s];
+      if (index != -1) {
+        *p = index;
```

Comment:
This point is we shouldn't have *two* kernels with a loop.  There are still 2 kernels with a loop.
The idea was, you could store the best arc index in an array with the same dimensions as the num-states for each FSA, in a location like `values[row_splits1[fsa_idx0] + num_arcs]`.  Then you could extract them in a kernel that doesn't have a loop.

Reply:
... I know in this particular case it's only a factor of 2, but this kind of thing is central about how the toolkit is intended to be used.  It'a all about removing the need for loops by using ragged tensors.

Reply:
I see. Will update it.

Reply:
... incidentally, it's possible to further optimize this so the run time is less than linear, by using methods similar to what I use in GetStateBatches(); imagine computing a table that instead of the best traceback, is the best 2-step traceback; and then the 4-step traceback, and so on.  There's no need for this right now, since the forward-pass algorithm is already at least linear-time in the length of this path.  But we can consider it for the future.

Reply:
Fixed. The second loop is eliminated and a TODO item is added to optimize it further as you suggested.